File: jcode-panel/jcode_panel/integrations/vscode.py
from __future__ import annotations

from pathlib import Path
import shutil

from .base import Integration, IntegrationStatus


class VSCodeIntegration(Integration):
    id = "vscode"
    name = "VS Code Context Extension"

    def __init__(self, project_root: Path):
        self.source_dir = project_root / "integrations" / "vscode_extension"
        self.target_dir = Path.home() / ".vscode" / "extensions" / "jcode-panel-context"

    def status(self) -> IntegrationStatus:
        installed = (self.target_dir / "package.json").exists()
        return IntegrationStatus(
            name=self.name,
            installed=installed,
            enabled=installed,
            message="Installed in ~/.vscode/extensions/jcode-panel-context. Reload VS Code to activate." if installed else "Not installed",
            install_hint="Install copies the extension locally. Reload VS Code, then move cursor in an editor before using @vscode.",
        )

    def install(self) -> IntegrationStatus:
        self.target_dir.parent.mkdir(parents=True, exist_ok=True)
        if self.target_dir.exists():
            shutil.rmtree(self.target_dir)
        shutil.copytree(self.source_dir, self.target_dir)
        return self.status()

    def uninstall(self) -> IntegrationStatus:
        if self.target_dir.exists():
            shutil.rmtree(self.target_dir)
        return self.status()

File: jcode-panel-tauri/src-tauri/src/integrations/vscode.rs
use serde::{Deserialize, Serialize};
use std::{fs, path::PathBuf, process::Command};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IntegrationStatus {
    pub installed: bool,
    pub message: String,
}

fn source_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../jcode-panel/integrations/vscode_extension")
}

fn target_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".vscode/extensions/jcode-panel-context")
}

pub fn status() -> IntegrationStatus {
    let installed_by_code = Command::new("code")
        .arg("--list-extensions")
        .output()
        .ok()
        .and_then(|out| String::from_utf8(out.stdout).ok())
        .is_some_and(|text| {
            text.lines()
                .any(|line| line == "jcode-panel.jcode-panel-context")
        });
    let installed = installed_by_code || target_dir().join("package.json").exists();
    IntegrationStatus {
        installed,
        message: if installed {
            "VS Code context extension installed"
        } else {
            "VS Code extension not installed"
        }
        .into(),
    }
}

pub fn install() -> IntegrationStatus {
    let src = source_dir();
    let dst = target_dir();
    if let Some(parent) = dst.parent() {
        let _ = fs::create_dir_all(parent);
    }
    if dst.exists() {
        let _ = fs::remove_dir_all(&dst);
    }
    if fs::create_dir_all(&dst).is_ok() {
        let _ = fs::copy(src.join("package.json"), dst.join("package.json"));
        let _ = fs::copy(src.join("extension.js"), dst.join("extension.js"));
    }
    status()
}
